### PR TITLE
Add Active Record 6.1 compatibility

### PR DESCRIPTION
**Pull request by tagliala**
- ActiveRecord Autoload Change: rails/rails@b744372
- Use DatabaseConfig Instead of Hash for Tasks: rails/rails#37291
- Move ActiveRecord::Tasks::DatabaseAlreadyExists to
  ActiveRecord::DatabaseAlreadyExists: rails/rails@69700c9
- Wrap native adapters connection errors: rails/rails#40110

Close: #322, #323

### CODE REVIEW

**Comment by keithdoggett** [comment]
@tagliala thanks for the PR. I was working on these changes after I saw your issues and it looks like our solutions are nearly identical. I'm going to post these links as references for the changes.

ActiveRecord Autoload Change: https://github.com/rails/rails/commit/b744372f2dfd3d86f7eb4af99b1f9049e21f3d44
Use DatabaseConfig Instead of Hash for Tasks: https://github.com/rails/rails/pull/37291/files
Move `ActiveRecord::Tasks::DatabaseAlreadyExists` to `ActiveRecord::DatabaseAlreadyExists`: https://github.com/rails/rails/commit/69700c9ee760c5880fc80af70a89b42bb791cf98 

I was able to pass all the tests on my branch. Were you able to get yours to work? I'll pull it later today and test it.

**Comment by tagliala** [comment]
Hi @keithdoggett , welcome

> Were you able to get yours to work? 

I was waiting for Travis CI, but the build didn't even start 😕

On my development machine, there are some failures related to containerized Postgres 12 with native pg 11 tools (same on master branch)

I would like to push another change (use symbols instead of strings for configuration variables), but feel free to close this or keep working on this commit, because I never worked on Active Record internals.

> ActiveRecord Autoload Change: rails/rails@b744372
> Use DatabaseConfig Instead of Hash for Tasks: https://github.com/rails/rails/pull/37291/files
> Move ActiveRecord::Tasks::DatabaseAlreadyExists to ActiveRecord::DatabaseAlreadyExists: rails/rails@69700c9

I think I should split the commit into three different ones, one for each change, and add a reference in the commit message

Let me know

File: lib/active_record/connection_adapters/postgis/create_connection.rb
Comment (tagliala) [inline lib/active_record/connection_adapters/postgis/create_connection.rb:37]: This should be done in a fourth commit. Changes has been made to keep uniformity with AR61's `PostgreSQLAdapter`

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (tagliala) [inline lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:40]: I would like to change `"database" =>` into `database:`

**Comment by keithdoggett** [comment]
We can just use this PR since ours are very similar.

Yes the Travis CI is very slow. I'm working on creating tests with GitHub Actions, so hopefully, soon I can get those up and running.

You don't need to worry too much about separating each change in ActiveRecord into its own commit here. I think as long as we reference the changes in ActiveRecord that's good enough.

I'll keep working on getting the GitHub Actions setup and then can pull this and see where ours differ and if yours works locally for me.

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (keithdoggett) [reply lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:40]: Yes I symbolized these keys on my version

**Comment by tagliala** [comment]
Ok, 

- symbolized keys
- changed commit message
- push forced

The previous Travis Build has been canceled, I hope that the new one can start faster

File: lib/active_record/connection_adapters/postgis/create_connection.rb
Comment (tagliala) [reply lib/active_record/connection_adapters/postgis/create_connection.rb:37]: It's the same as rails/rails#37291

Comment (tagliala) [reply lib/active_record/connection_adapters/postgis/create_connection.rb:37]: Found it. We need another change

rails/rails#40110

**Comment by tagliala** [comment]
Meanwhile, I'm testing this branch on my application

✅ https://travis-ci.org/github/diowa/icare/builds/749268657

The build was failing before Travis was not able to create the schema

**Comment by keithdoggett** [comment]
Are you able to run the tests locally with `rake test`?

**Comment by tagliala** [comment]
> Are you able to run the tests locally with `rake test`?

```
$ rake
The dependency activerecord-jdbcpostgresql-adapter (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The dependency ffi-geos (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
Run options: --seed 41095

# Running:

...........Created database 'postgis_tasks_test'
.Created database 'postgis_tasks_test'
...................................................

Fabulous run in 73.050773s, 0.8624 runs/s, 2.7515 assertions/s.

63 runs, 201 assertions, 0 failures, 0 errors, 0 skips
```

File: test/tasks_test.rb
Comment (tagliala) [inline test/tasks_test.rb:14]: `configuration_hash` in DatabaseConfigs is frozen

**Comment by tagliala** [comment]
Travis is running an older hash 🤦🏻‍♂️

Will push-force when it is finished to avoid conflicts

**Comment by keithdoggett** [comment]
@tagliala tomorrow I'll be able to get a GitHub workflow up that can replace the travis tests. Once I push that, you can merge that into your branch and we can validate off of that.

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (tagliala) [inline lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:103]: This is not possible anymore, because `configuration_hash` is frozen

Ref: https://travis-ci.org/github/rgeo/activerecord-postgis-adapter/builds/749304801

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (tagliala) [inline lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:103]: frozen

**Comment by keithdoggett** [comment]
I was able to get the testing workflows setup on this branch https://github.com/rgeo/activerecord-postgis-adapter/tree/github-workflows. Can you please merge it into this branch and we can test against that. I left both the `ar60` and `ar61` Gemfiles in there, but we will have to remove the `ar60` Gemfile because this change is not backward compatible.

In the meantime, I'll compare your changes to mine and start a review.

**Comment by tagliala** [comment]
muuuuch faster :)

File: activerecord-postgis-adapter.gemspec
Comment (keithdoggett) [inline activerecord-postgis-adapter.gemspec:23]: I think we should just pin this to `>= 6.2.2`. Not sure that it's necessary to specify both here.

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (keithdoggett) [reply lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:103]: We should freeze the resultant hash since this assignment undoes it.

```rb
h1 = {a: 1, b: 2}.freeze
p h1.frozen?
#=> true

h2 = h1.merge({c:3})
p h2.frozen?
#=> false
```

It is probably not a big deal, but I'd rather not have this be the source of weird downstream errors

**Review by keithdoggett — CHANGES_REQUESTED** [review]
A few minor changes are needed, but this looks good overall and is pretty much ready. Thanks for the PR!

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (tagliala) [reply lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb]: Good point. Added `.freeze`

**Comment by tagliala** [comment]
Any chance to merge the github-actions branch to master so I can rebase?

**Comment by keithdoggett** [comment]
I was planning on putting this into an `ar61` branch for a little bit since I have a few breaking changes I'm going to release with this. Does it work if I merge it there?

**Review by keithdoggett — APPROVED** [review]

**Comment by tagliala** [comment]
Yes, I will rebase against that branch and change the PR target

**Comment by keithdoggett** [comment]
I merged the `github-workflows` branch with `ar61`.

**Comment by keithdoggett** [comment]
This should be good to go. I'll let @BuonOmo review it when he gets the chance and then I'll merge it in. Thanks for the help!

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (BuonOmo) [inline lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:68]: A more correct pattern for using default values in ruby is
```suggestion
          @su_username ||= configuration_hash.fetch(:su_username, username)
```

This is more resilient to possible falsy values, and in terms of code readability, is more expressive IMHO. Could do the change here and for other occurences ?

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (BuonOmo) [inline lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:77]: I find this pattern a little bit clearer, but may not be that important

```suggestion
          @has_su if defined?(@has_su)
          
          @has_su = configuration_hash.include?(:su_username)
```

**Review by BuonOmo — COMMENTED** [review]
Thanks for the PR! I've made two comments, that I think will make the code a bit more readable.

Otherwise, LGTM!

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (schmidt) [reply lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:77]: Sorry for jumping in here. But I think, there's a bug in the suggested change. It should be

```diff
- @has_su if defined(@has_su)
+ return @has_su if defined(@has_su)
```

otherwise, the value for `@has_su` would be computed repeatedly.

Comment (BuonOmo) [reply lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb:77]: @schmidt of course, edited!

File: lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb
Comment (tagliala) [reply lib/active_record/connection_adapters/postgis/postgis_database_tasks.rb]: Done for both `su_username` and `su_password`. I can't see other occurrences where this apply in this file

**Review by BuonOmo — APPROVED** [review]
LGTM 🙂

**Comment by keithdoggett** [comment]
Merged :tada: thanks @tagliala!

I have a few breaking changes I've been working on with one that depends on the release of rgeo-activerecord v7, but I'm hoping to have ar61 merged into master within a week.

**Comment by tagliala** [comment]
Thanks to you!

**Pull request merged**
